Rewrite DAS Data Sort script into Object Orientated Class

nix-chu opened the pull request:
## Description
- Purpose of script is to convert a group of data points into one singular data point, based on the time unit provided by user. 

1. Reads the given CSV file and the data columns.
2. Combines the milliseconds within the given time intervals (seconds/minutes) into a singular time unit.
3. Then, the corresponding data points are averaged within each time interval. Excludes gps data, which will return a zero if the gps was turned off at this time, or a one if the gps was turned on. 
4. Creates a new CSV file and prints the newly manipulated data.

## Steps to Test
Only CSV files made after Ford 2019 will work with this script. I am currently using data_291.csv to test my script, which can be found on this thread. 
1. In the utils folder, transfer one of the CSV data files from the DAS into this folder. 
2. In the terminal, run `python das_data_sorter --input (file name) --output (output name) --unit (seconds/minutes)`
3. When finished, it should print 'Success!' into the terminal.

nix-chu committed:
Add Sorting Data script

nix-chu committed:
Update .gitignore file with .xlsx extension

nix-chu committed:
Create loop for indexes sorting into seconds

nix-chu committed:
Implement averaging for rider stats

nix-chu committed:
Import newly-manipulated data onto new CSV files

nix-chu committed:
Change Reed and Time variable names

nix-chu committed:
Fix formatting errors and commenting

nix-chu committed:
Fix formatting errors and commenting 2

nix-chu committed:
Change 'time' to 'milliseconds'

nix-chu committed:
Add averaging into minutes

nix-chu committed:
From past PR, add required parameter in parser

nix-chu committed:
Add choices parameter in parser, and remove else code block in 'time' if statement

nix-chu committed:
Fix parser code format to fit in code editor screen

nix-chu committed:
Merge branch 'master' into nix.chu/sorts-data-by-minutes

nix-chu committed:
Add pipfiles

nix-chu committed:
Merge branch 'nix.chu/sorts-data-by-minutes' of https://github.com/monash-human-power/data-acquisition-system into nix.chu/sorts-data-by-minutes

nix-chu committed:
Change i and t variables to more meaningful

nix-chu committed:
Rearrange functions to top of script

nix-chu committed:
Add Pipfiles

nix-chu committed:
Implement mean_smooth and median_smooth functions

nix-chu committed:
Adapt avg_data to utilise Statistics module

nix-chu committed:
Integrate existing code with smoothing function

nix-chu committed:
Update pipfile

nix-chu committed:
Merge branch 'master' into nix-chu/smooths_data_ver1

nix-chu commented:
For lines 12 and 27, let me know if rounding on another line is unconventional. I did this because I thought line 11 and 26 looked messy enough.

Also, let me know if rounding to two decimal places is also sufficient.

khanguslee commented:
BTW, you can reference other PRs by putting the full link of the PR here OR (if it's the same repo, use #<number>) see https://help.github.com/en/github/writing-on-github/autolinked-references-and-urls for an example.

khanguslee commented:
Next time, you can add a comment to a line in your code. So it's easier for the reviewer to look at these lines of code!

khanguslee commented on utils/das_data_sorter.py:
```diff
@@ -1,6 +1,79 @@
+from numpy import median, mean
+
+# averages data within the seconds or minutes, according to new_time array
+def avg_Data(array):
```
Was it intentional to have two functions called `avg_Data` ? It also looks like theres a small difference in both of them too, so which one did you want?

Also forgot to say this last time, can you rename this method to `average_data`. See https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables on how to name methods using the PEP 8 python conventions.

khanguslee commented on utils/das_data_sorter.py:
```diff
@@ -1,6 +1,79 @@
+        third = data[i+1]
+
+        new_data_point = mean([first, second, third])
+        new_data_point = round(new_data_point, ndigits=2)
```
I'd like to control the number of digits to round. Can you add a new `precision=2` argument to the `mean_smooth` function (By adding `=2`, we set a default parameter if the user does not include a second parameter to this function.

khanguslee commented on utils/das_data_sorter.py:
```diff
@@ -1,6 +1,79 @@
+def smooth_data(data, technique):
+    if technique == "mean":
+        # when input is "--smooth mean"
+        final_data = {
```
I think I would have preferred if you had iterated through the keys and values of the `data` dictionary and then added it to the `final_data` dictionary. This is because the keys for both dictionaries are the same.

nix-chu replied:
How do I specify the names for the keys (which represent each column of data without hardcoding it?

khanguslee commented on utils/das_data_sorter.py:
```diff
@@ -1,6 +1,79 @@
+
+# smoothing technique function caller
+def smooth_data(data, technique):
+    if technique == "mean":
```
The if statements are quite unnecessarily long. When I look at the two if statements, I can see that they are very similar. The only difference is that we use a different calculation method depending on the technique. Try assigning the technique function to a variable so it is cleaner.

Here is an example of how I would resolve this problem.

``` python
smoothing_method = None
if technique is "mean":
	smoothing_method = mean_smooth
elif technique is "median":
	smoothing_method = median_smooth
else:
	raise ValueError("Invalid technique given")

final_data = {
            "time": range(1, len(new_time)-1),
            "gps_course": smoothing_method(data["gps_course"]),
            "gps_speed": smoothing_method(data["gps_speed"]),
            "tempC": smoothing_method(data["tempC"]),
            "tempF": smoothing_method(data["tempF"]),
            "cadence": smoothing_method(data["cadence"]),
            "power": smoothing_method(data["power"]),
            "reed_velocity": smoothing_method(data["reed_velocity"]),
            "reed_distance": smoothing_method(data["reed_distance"])
        }
return final_data
```

hallgchris replied:
In general when coding, if you've ever had to type the same thing twice, there's almost always a better way :wink:

khanguslee commented on utils/das_data_sorter.py:
```diff
@@ -21,6 +94,9 @@ def avg_Data(array):
 parser.add_argument("--unit", help="Specify time units (seconds, s, or minutes, m)", default="seconds", 
                     choices=["seconds", "s", "minutes", "m"], action="store")
+
+parser.add_argument("--smooth", help="Smoothens data points by 3-point mean or median smoothing", 
```
What happens if a user does not input a `--smooth` argument?

nix-chu replied:
I suppose in this case, the user would just want normal converted data. Some users might want this, so that they can know the exact values at certain points of time.

khanguslee commented on utils/das_data_sorter.py:
```diff
@@ -72,5 +148,15 @@ def avg_Data(array):
+
+# checks if user wants to smooth data using moving mean or moving median method
+# if no input from user, script will ignore this whole block
+if args.smooth == "mean":
```
Seems like a unnecessary if statement here? We can simply just pass in `args.smooth`

khanguslee reviewed (changes requested):
Some changes to make, but very minor! This will come in time as you continue programming more and more!

hallgchris commented on utils/das_data_sorter.py:
```diff
@@ -1,6 +1,79 @@
+
+    return avg_array
+
+# 3 point moving mean smoothing
```
When you have a comment which is documenting a function, we usually use python's multiline comments, i.e.
```python
def mean_smooth(data):
""" 3 point moving mean smoothing """
    new_data = [] # etc.
```
If you do your function documentation like this, whatever editor will recognise it and display it as the description of a function when you mouse over it (for example)
![image](https://user-images.githubusercontent.com/17876556/79969695-2e6cbc80-84d5-11ea-8417-a8a0fa7128c2.png)

hallgchris commented on utils/das_data_sorter.py:
```diff
@@ -1,6 +1,79 @@
+        second = data[i]
+        third = data[i+1]
+
+        new_data_point = median([first, second, third])
```
Do we not want to round when we're calculating medians?

hallgchris commented on utils/das_data_sorter.py:
```diff
@@ -1,6 +1,79 @@
+def avg_Data(array):
+    avg_array = []
+
+    for millisecond_indices in new_time:
```
Here we're accessing the variable `new_time`, which isn't defined until much later in the file. Python allows this, but in most cases I'd try to avoid accessing variables outside of the scope of the function.

The reason for this is mainly because it makes code easier to read, and easier to debug. When I first read this code, I had no clue what this variable was, but I would have a better idea if it was instead a parameter of the function. As for debugging, functions are much easier to debug when they are deterministic - meaning that for the same inputs, they will always return the same outputs. This is not the case here, as the function's output also depends on the value of `new_time` which is completely independent.

`new_time` appears in a few other functions around this file, they could be made parameters too!

hallgchris reviewed (changes requested):
Just a few things to add on to Angus's feedback! On the whole looking really good though

nix-chu committed:
Rewrite script to Object-Orientated Class

nix-chu committed:
Fix formatting errors for class methods

nix-chu committed:
Implement mean function

nix-chu committed:
Change file name

nix-chu committed:
Delete das_data_sorter2.py

Duplicate file. 

Also I don't know how to delete files and then pushing that change using the terminal.

hallgchris commented on utils/Pipfile:
```diff
@@ -9,6 +9,7 @@ verify_ssl = true
 pandas = "*"
 argparse = "*"
 math = "*"
```
What is the `math` package, is it meant to be the one built into python? It's causing issues for me trying to `pipenv install`. I think this line needs to be removed.

nix-chu replied:
@hallgchris Oops, forgot about the pipfile. I didn't change this file and remove the math package because I didn't know what to do with it, since we are moving on to poetry. Should I delete this file instead?

hallgchris commented on utils/das_data_sorter.py:
```diff
@@ -1,76 +1,170 @@
+# accepts terminal arguments
+parser = ArgumentParser()
 parser.add_argument("--file", help="Input CSV file", action="store", required=True)
 parser.add_argument("--output", help="Returns the filtered data", default="filtered_data.csv", action="store")
 parser.add_argument("--unit", help="Specify time units (seconds, s, or minutes, m)", default="seconds", 
                     choices=["seconds", "s", "minutes", "m"], action="store")
```
`output` and `unit` are probably required fields, like `file`.

nix-chu replied:
I didn't make these required because I have put a default argument for these. 
Output default = `filtered_data.csv`
Unit default = `seconds`
Is this fine or still change it?

hallgchris replied:
Oh yeah that's fine then :)

hallgchris replied:
Actually - I don't think we should have a default output, as it could lead to accidentally overwriting a file. We should either have no default, or have some confirmation about whether you really want to overwrite. (The former would be much easier)

hallgchris commented on utils/das_data_sorter.py:
```diff
@@ -1,76 +1,170 @@
-parser = argparse.ArgumentParser()
+# accepts terminal arguments
+parser = ArgumentParser()
 parser.add_argument("--file", help="Input CSV file", action="store", required=True)
```
Should call this argument `input`, just for consistency?

khanguslee replied:
yes pls

hallgchris commented on utils/das_data_sorter.py:
```diff
@@ -1,76 +1,170 @@
+        
+        args.unit accepts only seconds/s and minutes/m as arguments. 
+        '''
+        if args.unit == "seconds" or args.unit == "s":
```
Ideally, we want to be able to use this `DasSort` class without needing to worry about what variables are set outside of it. Here, we're referencing a variable `args` which might not necessarily exist if we for example created an instance of this class from another python file. (Another example - someone might change `args` without updating `DasSort`, as there's no clear connection between the two unless you read all the code)

Instead, I would pass the different bits of `args` that you need in the constructor of `DasData` and store them to be used whenever, such as here. That way, the variables have a clear obvious association with the class and will always be accessible, regardless of what happens outside the class.

hallgchris commented on utils/das_data_sorter.py:
```diff
@@ -1,76 +1,170 @@
-})
-final.to_csv(args.output, index=False)
+class DasSort:
+    def __init__(self, data:str) -> None:
```
Awesome to see type hinting! The type of `data` should be `pd.DataFrame`, I believe.

![image](https://user-images.githubusercontent.com/17876556/81806994-9bb5bf80-9560-11ea-86fc-e79102202ed0.png)

hallgchris commented on utils/das_data_sorter.py:
```diff
@@ -1,76 +1,170 @@
+        args.unit accepts only seconds/s and minutes/m as arguments. 
+        '''
+        if args.unit == "seconds" or args.unit == "s":
+            milliseconds = milliseconds/1000
```
I think we could have a better name for this variable, as the data is no longer in units of milliseconds.

hallgchris commented on utils/das_data_sorter.py:
```diff
@@ -1,76 +1,170 @@
+                # None and zeroes are ignored
+                length -= 1
+                continue
+            elif number > 0 or number < 0:
```
Unless you've thought of some special case I haven't thought of, if `number == 0` was false, then `number > 0 or number < 0` will always be true. So, this case in the if statement will always run, and the `else` section below will never be reached. You should be able to replace the whole thing with `total += number`

nix-chu replied:
I had a problem earlier where, in the CSV file, if the cell was empty, it would say that I was trying to perform addition with an integer and a string. (Shown below is an outdated screenshot)
![image](https://user-images.githubusercontent.com/58575113/82114274-15d38780-979f-11ea-98cc-940bcf130ef2.png)

So I thought that an empty cell in Python would read as "". However, after some debugging, I found out that an empty cell's data type is `float`, so I'm just all-round confused. Hence why `pd.isna(number)` is there, which is Pandas's way of handling empty cells. I also kept the else code block there for future debugging. 

(Shown below is the error message when elif statement has `number == ""`)
![image](https://user-images.githubusercontent.com/58575113/82113798-128acc80-979c-11ea-9331-14706e785bfe.png)

The link below is a reference I found on how to handle empty cells in Python, scroll all the way down to the "Dealing with nan without getting a headache" header:
https://towardsdatascience.com/navigating-the-hell-of-nans-in-python-71b12558895b

hallgchris commented on utils/das_data_sorter.py:
```diff
@@ -1,76 +1,170 @@
+
+        return universal_array
+
+    def mean(self, data_array:list) -> int:
```
Input type is pd.Series, return type will be `float`.

Although, `list`s will still work in this function. If you wanted to be really fancy you could do `Union(pd.Series, list)` from the typing module, but that isn't necessary as we don't pass this function `list`s from anywhere.

hallgchris commented on utils/das_data_sorter.py:
```diff
@@ -1,76 +1,170 @@
+        '''
+        new_location_data = []
+        for index_array in self.indexes:
+            first_element = index_array[0] 
```
`first_element_index` would be more descriptive

hallgchris commented on utils/das_data_sorter.py:
```diff
@@ -1,76 +1,170 @@
+            "reed_distance": self.reed_distance
+        })
+        final.to_csv(args.output, index=False)
+        print("Success! Check in the same folder for the new CSV file.")
```
The output file isn't located in the same folder as the input, it's located wherever the user specified. Maybe instead print `Success! Output written to {args.output}`

khanguslee replied:
I was about to say this!

hallgchris commented on utils/das_data_sorter.py:
```diff
@@ -1,76 +1,170 @@
+        
+        return new_location_data
+    
+    def print_output_file(self) -> None:
```
We "print" something when it appears on the console, but when we're putting stuff into a file, "write" might be more clear

khanguslee replied:
As a reader of this code, I thought this function would print the whole output file. But once I looked at the function contents, it does something else. Try describe this function in a few words about what it does. That should help you in naming this function!

hallgchris commented on utils/das_data_sorter.py:
```diff
@@ -1,76 +1,170 @@
+    def __init__(self, data:str) -> None:
+        self.indexes = None
+        self.time = self.convert_time(data["time"])
+        self.gps = self.average_data(data["gps"])
```
As far as I know, `gps` contains either a 1 when gps data is available, and 0 otherwise. Rather than averaging this data, I think it would make most sense to have a 1 if there's any 1s in the original data, and only a 0 if all values are 0s.

nix-chu replied:
I understand what you are saying, and it would make the script run quicker if it didn't use the average_data method. One problem is that I'm not sure how to go about this. 

When I tried checking if there was a 1 within the time interval, by doing `if 1 in data[index_array]`, it didn't work the way it's supposed to work. After doing some testing on a different Python file, I found that the data points inside the gps column is **not** of the integer type
![image](https://user-images.githubusercontent.com/58575113/82115789-4c61d000-97a8-11ea-9856-49ccdf4127e8.png)

hallgchris replied:
That's odd, it seems to be working for me
![image](https://user-images.githubusercontent.com/17876556/82132286-8b803780-9821-11ea-94d3-74052297a165.png)

hallgchris reviewed (changes requested):
Some great work here! Don't forget to mark the PR ready for review if you've made it a draft or tagged it as WIP. And sorry for taking a few days :)

khanguslee commented on utils/das_data_sorter.py:
```diff
@@ -1,76 +1,170 @@
-# averages data within the seconds, according to seconds array 
-def avg_Data(array):
-    avg_array = []
+D = TypeVar('DAS_data') # data type to represent columns of the data in a given file
```
I really like the concept of this! Can we change the name `D` into a more descriptive type?

nix-chu replied:
I got rid of this hahaha. I used this to represent columns of data points, but turns out they have their own data type as discovered by Chris. 

Just copied this concept from my FIT1008 classes, but still unsure how to fully use this. Might ask my lecturer before utilising this.

nix-chu replied:
So according to my lecturer, using TypeVar is like using pronumerals in maths. For a math function like; f: R->R, f(x)=x^2+x+3, this function only allows real numbers to substitute for x. But in Python, TypeVar makes it so that x must be of the same data type throughout the script. This is to allow flexibility in what data types the user inputs.

If the Python function is `def function(arg1: T, arg2 :T)` then that means the argument passed by both arg1 and arg2 must both be strings. Or in another case, both must be integers. Depends on how the user wants to use the class. This ensures that the user is using the same data types for the class. 

For example, we cant do `function("str", 48)`, because T is confused whether it is a string or an integer. But we can do `function("str", "str")` or `function(48, 32)`. 

Upon saying that, she suggests that you only use TypeVar if the type to be used for all scripts is unknown. Otherwise, there's no point in using TypeVar if the arguments passed are intended for a specific data type, because you can just do `def function(arg1: int, arg2: int)`.

khanguslee commented on utils/das_data_sorter.py:
```diff
@@ -1,76 +1,170 @@
+        self.gps_course = self.average_data(data["gps_course"])
+        self.gps_speed = self.average_data(data["gps_speed"])
+        self.gps_satellites = self.average_data(data["gps_satellites"])
+        self.ax = self.average_data(data["aX"])
```
Once we eventually add `pylint`, it will complain about this class because there's too many variables. This is fine for now but we might want to make some helper classes to store these variables instead.

nix-chu replied:
Not sure how to do that, I haven't heard of a helper class before

khanguslee commented on utils/das_data_sorter.py:
```diff
@@ -1,76 +1,170 @@
+        avg_data = []
+
+        for index_array in self.indexes:
+            avg = self.mean(data[index_array])
```
I like to name this variable to something like `current_average`. Thoughts?

khanguslee commented on utils/das_data_sorter.py:
```diff
@@ -1,76 +1,170 @@
+            # in the event where length = 0, due to all the elements in data_array being ignored
+            return 0 
+
+    def average_data(self, data:D) -> D:
```
I just realised it looks like theres a smiley face and a shocked face `:D) -> D:`

hallgchris replied:
The function takes a happy face and makes it sad :(

khanguslee replied:
@hallgchris why is it `average_data`? Why can't it be `happy_data`?

khanguslee reviewed (changes requested):
The code looks so much better than what you had before! Nearly there!

nix-chu committed:
Make changes on Python file

nix-chu committed:
Merge branch 'nix-chu/smooths_data_ver1' of https://github.com/monash-human-power/data-acquisition-system into nix-chu/smooths_data_ver1

khanguslee reviewed (dismissed):
LGTM. Keen to see what @hallgchris says!

hallgchris commented on utils/das_data_sorter.py:
```diff
@@ -162,9 +171,9 @@ def print_output_file(self) -> None:
-        final.to_csv(args.output, index=False)
-        print("Success! Check in the same folder for the new CSV file.")
+        final.to_csv(file_output, index=False)
+        print("Success! Output is written to {file_output}")
```
You'll need to use backticks (``` ` ` ```) rather than quotes to use python's f-strings. (have a google if you're unfamiliar, they're super useful)

nix-chu replied:
Can't seem to use backticks, but doing `print(f"Success {file_output}")` seems to work well ¯\\\_(ツ)\_/¯ ahahaha

hallgchris replied:
Sorry lol I'm confusing python with JS :sweat_smile: That's great now

hallgchris reviewed (changes requested):
I've just realised an issue - Sometime since ford last year, we replaced the `gps_location` column with three columns, `gps_lat`, `gps_long` and `gps_alt`. As a result, this script crashes when trying to process a log made since around the middle of last year. I think what we need is for this script to see if the `gps_location` column exists, and if it does, process that column, otherwise process the newer three columns. (should also document why you do this, as it may not be clear looking back on the code).

GitHub doesn't seem to let me upload a `.csv`, so I'll send you a log via slack DM

hallgchris commented:
Nevermind, when I tried to upload it outside of the review environment I got a better error message, just needed to be zipped to make it happy. This is from Al's run from Holden this year :)

[data_291.zip](https://github.com/monash-human-power/data-acquisition-system/files/4639509/data_291.zip)

nix-chu committed:
Remove Pipfiles

nix-chu committed:
Use f-strings to print success statement

khanguslee commented on utils/das_data_sorter.py:
```diff
@@ -1,76 +1,179 @@
+        print("Success! Output is written to {file_output}")
+
+if __name__ == '__main__':
+    data = pd.read_csv(args.input)
```
This looks so much neater than the old script! I can now read at a high-level as to what and how this program works.

khanguslee commented on utils/das_data_sorter.py:
```diff
@@ -1,76 +1,179 @@
+        self.reed_velocity = self.average_data(file_input["reed_velocity"])
+        self.reed_distance = self.average_data(file_input["reed_distance"])
+
+        self.write_to_output_file(file_output)
```
This should not be ran inside the class instructor. This should be ran by the class instance eg. `das_sort.write_to_output_file(args.output)`. If you do this, you also don't need to include the `file_output` argument.

khanguslee commented on utils/das_data_sorter.py:
```diff
@@ -1,76 +1,179 @@
+
+if __name__ == '__main__':
+    data = pd.read_csv(args.input)
+    das_sort = DasSort(data, args.output, args.unit)
```
Once you add the above changes. This will basically read as:
1. Load csv file
2. Sort csv file
3. Write csv file to a file

If a developer wants to know _how_ the data is sorted, they will read `DasSort` implementation. I hope this gives you an idea about why we would do this!

khanguslee reviewed (changes requested):
Some small changes that I also did not pick up last time oops. It's a really small change!

khanguslee commented:
Also, your commit messages are really good btw! You really got the hang of it! Just get into the habit of committing with one feature but I don't always follow that too ;)

nix-chu committed:
Fix gps_data code

nix-chu committed:
Move Write Output File method outside constructor

nix-chu committed:
Replace gps_location into gps_lat, gps_long, gps_alt

nix-chu commented:
Some very vital changes made in the last few commits, so I will summarise them so that each key change can be reviewed. I also think that it'd be a good idea to discuss these and gather opinions about the code changes that I made:
1. Was able to fix gps_data method. 

In pandas, for some reason `if 1 in data[index_array]` (line 132) takes the indexes. However `if 1 in data[index_array].values` works, because it was accessing the actual array of values. So this returns whether the GPS was off (0) or on (1) at each time interval, which is what it's intended to do 👍

2. Moved the write_to_output_file outside the constructor, previously it was the last line in `__init__`. 

Angus' logic makes sense, this allows the user to take each step in the process one-by-one. Firstly, read the CSV file, then filter the data, and finally write the filtered data onto a new file. Aside from this, we can adapt the code in the future to allow the user to manipulate the data before writing it onto a new file.

My initial idea for this script was to automate everything for the user, but this is better because it allows flexibility in how the user wants the data to be _transformed_.

3. Replaced gps_location into gps_lat, gps_long and gps_alt.

As Chris mentioned, the new CSV files (from Ford 2019) contain gps_lat, gps_long and gps_alt, instead of gps_location. I wasn't sure if it's possible to have an if statement inside the constructor to check if the CSV file has gps_location OR gps_lat/long/alt. So I decided that it might be better to simply have gps_lat/long/alt as the instance variables. Since gps_location will no longer be a column in future CSV files, then the chunks of code written for handling gps_location will be eventually redundant. 

This will simply mean that the script will only work for CSV files AFTER Ford 2019. (If the columns are the same)

That's all the key changes I've made. Let me know if you have a second opinion/idea for the key changes I've made and other feedback. Thanks!

nix-chu committed:
Remove numpy.int64 method as it is no longer needed

khanguslee commented:
> 1. Moved the write_to_output_file outside the constructor, previously it was the last line in `__init__`.
> My initial idea for this script was to automate everything for the user, but this is better because it allows flexibility in how the user wants the data to be _t r a n s f o r m e d_.

- Your logic makes sense if we wanted to write a script that only does one thing. But if we want to think about it long term, we can reuse your class in different ways!

> 1. Replaced gps_location into gps_lat, gps_long and gps_alt.
> 
> This will simply mean that the script will only work for CSV files AFTER Ford 2019. (If the columns are the same)

- Yeah oops, this is my fault. Perhaps we can create some sort of converter that just converts V2 to V3 logs one day..

khanguslee reviewed (dismissed):
LGTM. Only gripe I have is docstrings! But we can sort that out when we have a standard we want in the DAS repo.

hallgchris commented:
> Yeah oops, this is my fault. Perhaps we can create some sort of converter that just converts V2 to V3 logs one day...

Agreed we should make a converter between these two formats, but this script isn't for V3 logs is it? Cos V3 logs have different columns depending on the sensor module

hallgchris commented on utils/das_data_sorter.py:
```diff
@@ -122,20 +122,22 @@ def average_data(self, data:pd.Series) -> pd.Series:
-    #     return new_gps_data
+        return new_gps_data
 
     def location_data(self, data:pd.Series) -> pd.Series:
```
Now that we're not dealing with the `gps_location` column, we can actually average the data, as it's no longer stored as a string!

hallgchris reviewed (changes requested):
Lucky last thing. All good after this!

nix-chu committed:
Removed location_data function, replaced with average_data function

nix-chu committed:
Install poetry

nix-chu committed:
Fixed missing gps_lat,gps_long,gps_alt columns on output CSV files

khanguslee commented:
Thoughts on having `poetry` inside the `utils` folder? It seems like `poetry` is located inside the `DAS` folder in https://github.com/monash-human-power/data-acquisition-system/pull/53. Should we just have a single poetry file at the root of the directory? Not sure what would be the best solution. @Blake-Haydon @hallgchris @nix-chu @angustrau.

hallgchris commented:
I reckon just have one single poetry file at the root per project. Unless one option is clearly better than the other (and do speak up if one is!), the simplest solution is always the best of two equal options

nix-chu commented:
Ah, I did this just to try installing poetry myself. I don't mind taking it out of utils folder if having a single poetry file works just fine :)

khanguslee commented:
@nix-chu Just leave your poetry file then. Would be cool if poetry had something like `yarn workspaces`.

nix-chu commented:
@hallgchris does you mean have one in my project and another in blake's etc? or just one for the entire repository?

hallgchris commented:
I was saying one for the entire repo, but thinking more about it, I'm happy to leave it as is. @Blake-Haydon's one is for stuff to be ran on the raspberry pi, yours is for stuff to run later on our computers, so I can see why they should be kept separate.

khanguslee commented:
Just wanted to note that, if you do a `poetry install` to install your dependencies for this project, it will create a `.egg-info` folder. See https://github.com/python-poetry/poetry/issues/800 for more information. To remove this, you will need to install libraries by doing `poetry install --no-root`. I think this will be fixed when we move to a single poetry file at the root of this project.
![Screen Shot 2020-05-25 at 3 21 36 pm](https://user-images.githubusercontent.com/18709969/82780739-6e8bda00-9e9b-11ea-8b54-78a9d8469edd.png)

khanguslee reviewed (approved):
Output from your script! (Don't forget these on your PR next time pls!)
<img width="1326" alt="Screen Shot 2020-05-25 at 3 38 55 pm" src="https://user-images.githubusercontent.com/18709969/82781704-d9d6ab80-9e9d-11ea-8cdc-6b1e5f05f8df.png">

Such a useful script!

hallgchris reviewed (approved):
Looks great!
